pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java

```diff
@@ -74,22 +74,16 @@ public synchronized T getValue() {
     public synchronized CompletableFuture<Void> updateValue(T newValue) {
         // If there is an operation in progress, we're going to let it complete before attempting to
         // update the value
-        if (pendingOperationFuture.isDone()) {
-            pendingOperationFuture = CompletableFuture.completedFuture(null);
-        }
-
-        pendingOperationFuture = pendingOperationFuture.thenCompose(v -> {
-            synchronized (ResourceLockImpl.this) {
-                if (state != State.Valid) {
-                    return CompletableFuture.failedFuture(
-                            new IllegalStateException("Lock was not in valid state: " + state));
-                }
-
-                return acquire(newValue);
-            }
-        });
-
-        return pendingOperationFuture;
+        return pendingOperationFuture = pendingOperationFuture.exceptionally(ex -> null) // ignore all the exception
```

Comment:
>I don't think we should ignore the exception here. If the pending revalidation fails, we are not the owner of the lock anymore and we shouldn't be updating the lock (which will anyway fail, because the version has changed).

Since `pendingOperationFuture ` is constantly updated by `silentRevalidateOnce`, that method is ensured the state will be released if we get a `LockBusy` or `BadVersion` exception. Therefore, in that case. This method will consistently return `IllegalState` instead of many different exceptions for the caller.

I am not sure if you can accept this way. If you don't think so. I can revert this part of the changes.

Reply:
>I don't think we should ignore the exception here. If the pending revalidation fails, we are not the owner of the lock anymore and we shouldn't be updating the lock (which will anyway fail because the version has changed).

Yes, you are right. Although `silentRevalidateOnce` can ensure the state will mark as `released` after the previous `revalidate` failed by `LockBusy` or `BadVersion`. But it can still have a chance to continue to revalidate by `updateValue` failed. Since the update failed doesn't mean the lock has expired. Therefore, I revert to original behaviour.

Reply:
>Another issue here is that we keep chaining at the back of the same future. With many updates the link of future can become very long. We should break it by starting with a dummy completed future when the existing one is completed.

There is a question:
From the source code of `CompletableFuture`, it will always create a new future while we use `thenCompose`. https://github.com/mattisonchao/jdk/blob/5b2e2e4695768a6bd8090fb9a6c342fcddcbb3fd/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L1170
So, what do you mean by "back of the same future. ". could you please explan it? : )

Reply:
@merlimat I have reverted this part of the code you mentioned. PTAL.

Reply:
> There is a question: From the source code of `CompletableFuture`, it will always create a new future while we use `thenCompose`. 

The compose will still be keeping a reference on the previous future (https://github.com/mattisonchao/jdk/blob/5b2e2e4695768a6bd8090fb9a6c342fcddcbb3fd/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L1173). Effectively it becomes a linked-list of futures.

Reply:
I see, thanks. @merlimat